compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributionPatterns.cpp

```diff
@@ -369,9 +370,161 @@ struct DistributeTransferWriteLayoutAttr final
     return accumulator;
   }
 };
+struct DistributeReductions final
+    : OpDistributionPattern<vector::MultiDimReductionOp> {
+  using OpDistributionPattern::OpDistributionPattern;
+
+  DistributeReductions(MLIRContext *context, int64_t maxBitsPerShuffle)
+      : OpDistributionPattern(context), maxBitsPerShuffle(maxBitsPerShuffle) {}
+
+  Value doThreadGlobalReduction(Value result, uint64_t shuffleOffset,
+                                int64_t laneSize,
+                                vector::CombiningKind combiningKind,
+                                int64_t entriesPerVector, Value mEmpty,
+                                OpBuilder &rewriter, Location loc) const {
+    uint32_t size = maxBitsPerShuffle;
+    Value mask;
+    for (uint64_t i = shuffleOffset; i < shuffleOffset * laneSize; i <<= 1) {
```

Comment:
There needs to be verification somewhere that `shuffleOffset` and `laneSize` are powers of 2. Probably best to do it when computing the offset and add an assert here.

Reply:
sg will add